pkg/controllers/cluster/infra.go

```diff
@@ -0,0 +1,97 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	awsv1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	azurev1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	gcpv1 "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/openshift/cluster-capi-operator/pkg/operatorstatus"
+)
+
+type GenericInfraClusterReconciler struct {
+	operatorstatus.ClusterOperatorStatusClient
+	InfraCluster client.Object
+}
+
+func (r *GenericInfraClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(r.InfraCluster).
+		Complete(r)
+}
+
+func (r *GenericInfraClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithName("InfraClusterController")
+
+	if err := r.Client.Get(ctx, req.NamespacedName, r.InfraCluster); err != nil && !errors.IsNotFound(err) {
```

Comment:
Make a copy of `r.InfraCluster`, don't use the one embedded in the object, it's not going to be threadsafe if you do that